README.md
# Cerberus Node Client

This is a node client for interacting with a Cerberus backend. It can be used in any environment that has [AWS credentials][node aws credentials] available.

To learn more about Cerberus, please visit the [Cerberus website](http://engineering.nike.com/cerberus/).

This library is compatible with node 12.x +, for previous versions of Node use the [1.x versions of the client](https://github.com/Nike-Inc/cerberus-node-client/tree/1.X_kms_client)

![Node.js CI](https://github.com/Nike-Inc/cerberus-node-client/workflows/Build/badge.svg)
[![codecov](https://codecov.io/gh/Nike-Inc/cerberus-node-client/branch/master/graph/badge.svg)](https://codecov.io/gh/Nike-Inc/cerberus-node-client) [![NPM][npm img]][npm] [![License][license img]][license]

## Installation

```
npm install --save cerberus-node-client
```

## Usage

See the [CerberusClient](http://engineering.nike.com/cerberus-node-client/CerberusClient.html) class on the [the docs site][docs]

## Authentication

The cerberus client uses the [AWS SDK Credentials provider chain][node aws credentials] to load AWS IAM credentials and authenticates with Cerberus via the [sts auth endpoint](https://github.com/Nike-Inc/cerberus-management-service/blob/master/API.md#app-login-sts-v2-v2authsts-identity)
This client will work in any environment that has access to AWS Credentials.

Cerberus will attempt to authenticate one its first call. The authentication result will be stored and reused. If the token has expired on a subsequent call, authentication will be repeated with the original configuration. You should not have to worry about authentication or token expiration; just use the client.

## Testing
First, make sure the following environment variables are set before running the Node Client integration tests:
```
export CERBERUS_HOST="https://example.cerberus.com"
export TEST_REGION="us-west-2"
export TEST_SDB="my-sdb"
export TEST_SDB_CATEGORY="app"
```
Ensure the TEST_SDB and TEST_SDB_CATEGORY variables match up with the path to your test sdb (i.e. app/my-sdb/test-path)

Then make sure AWS credentials have been obtained. One method is by running [gimme-aws-creds](https://github.com/Nike-Inc/gimme-aws-creds):
```
gimme-aws-creds
```

Next, in the project directory run: 
```
npm run style && npm run test:unit:local && npm run test:integration
```

If applicable, we recommend running the Node Client integration tests in both AWS Global and AWS China using the corresponding hosts, regions, and AWS credentials.

## A Note about Lambdas and Cerberus

While this client supports any env with IAM credentials, generally it does NOT make sense to store Lambda secrets in Cerberus for two reasons:

1. Cerberus can't support the scale that lambdas may need, e.g. thousands of requests per second
1. Lambdas will not want the extra latency needed to authenticate and read from Cerberus

A better solution for Lambda secrets is using the [encrypted environmental variables](http://docs.aws.amazon.com/lambda/latest/dg/env_variables.html)
feature provided by AWS.

Another option is to store Lambda secrets in Cerberus but only read them at Lambda deploy time, then storing them as encrypted
environmental variables, to avoid the extra Cerberus runtime latency.

## License

Cerberus Management Service is released under the [Apache License, Version 2.0](http://www.apache.org/licenses/LICENSE-2.0)

[travis]:https://travis-ci.org/Nike-Inc/cerberus-node-client
[travis img]:https://api.travis-ci.org/Nike-Inc/cerberus-node-client.svg?branch=master

[license]:https://github.com/Nike-Inc/cerberus-node-client/blob/master/LICENSE.txt
[license img]:https://img.shields.io/badge/License-Apache%202-blue.svg

[npm]:https://www.npmjs.com/package/cerberus-node-client
[npm img]:https://img.shields.io/npm/v/cerberus-node-client.svg


[coverage img][![codecov](https://codecov.io/gh/Nike-Inc/cerberus-node-client/branch/master/graph/badge.svg)](https://codecov.io/gh/Nike-Inc/cerberus-node-client)

[docs]:http://engineering.nike.com/cerberus-node-client/
[node aws credentials]:https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/setting-credentials-node.html


CHANGELOG.md
# Change Log /  Release Notes
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [4.0.0] - 2022-11-09
## Deprecated
- Node 12 and below are no longer supported
## Fixed
- Updated all dependencies to latest versions
## Added

## [3.0.0] - 2020-06-25
## Deprecated
- Node 8 is no longer supported
## Fixed
- Updated all dependencies to latest versions
- Fixed standard formatting errors
- Improved test coverage
## Added
- Added tests for logger
- Compatibility to run cerberus-node-client within Kubernetes

## [2.2.4] - 2020-06-25
## Fixed
- Updated version of minimist

## [2.2.3] - 2020-06-18
## Added
- Add typescript definition to published package

## [2.2.2] - 2020-06-16
## Fixed
- Use correct NPM OTP

## [2.2.1] - 2020-06-16
## Fixed
- Use correct Travis credentials

## [2.2.0] - 2020-06-16
## Added
- Typescript declaration

## [2.0.1] - 2019-06-20
## Fixed
- Use Client ctor `token` option

## [2.0.0] - 2018-12-19
## Added
- Support for STS Authentication
- JSDocs
## Removed
- Support for less than Node 8
- All logic relating to KMS authentication
- All logic that read user credentials from a TTY terminal

## [1.6.0] - 2018-07-16
## Added
- Support for File API from @mayitbeegh

## [1.5.2] - 2018-06-18
## Fixed
- Removed sensitive log info

## [1.5.1] - 2018-06-15
## Fixed
- Resolved issue with missing credentials in ECS

## [1.5.0] - 2018-06-08
## Added
- AWS ECS support

## [1.4.0] - 2018-06-04
## Added
- Authentication with assume role

## [1.3.0] - 2017-10-18
## Changed
- Use Cerberus v2 IAM principal auth. The previous version (v1) is deprecated.

## [1.1.2] - 2017-06-14
## Fixed
Error log from getting lambda configuruation

## [1.1.1] - 2017-06-14
## Added
Additional error guard to KMS

## [1.1.0] - 2017-06-07
## Added
Cerberus client header
## Changed
Coverage tool to Istanbul

## [1.0.0] - 2017-05-19
## Changed
- Required node version to v4.x
- Using promises internally
## Removed
- AWS SDK dependency

## [0.3.3] - 2017-05-08
### Changed
- Increase token expiration padding from 10 to 60 seconds

## [0.3.2] - 2017-05-03
### Changed
- Log key result after error check

## [0.3.1] - 2017-04-05
### Fixed
- `setLambdaContext` bug introduced in 0.3.0

## [0.3.0] - 2017-04-05
### Added
- MFA Support for prompt login

## [0.2.5] - 2017-03-31
### Changed
- Check for Cerberus errors before checking status code

## [0.2.4] - 2017-03-31
### Added
- Check for non-200 status from Cerberus Key Request

## [0.2.3] - 2017-03-31
### Added
- Return error from Cerberus for auth calls
- Tests for empty and error cerberus response

## [0.2.2] - 2017-03-06
### Fixed
- Token expiration logic

## [0.2.1] - 2017-01-06
### Changed
- Error check for cerberus server response

## [0.2.0] - 2016-11-15
### Added
- Code coverage npm script
- CHANGELOG.md
- CONTRIBUTING.md
- LICENSE.md
- User credentials (env and prompt) flow

### Changed
- Moved `test`
- split npm `test` script to into `style`, `unit`.
- Updated npm `test` script to wrap `style`, `unit`, and new `coverage` script
- Use Apache 2.0 License
- Start versioning at 0.2.0 to represent minor increase from pre-OSS version
- Added Tim Kye as maintainer in README


index.js
'use strict'
const request = require('request-micro')
const backoff = require('exponential-backoff')
const urlJoin = require('path').posix.join
const FormData = require('form-data')
const packageData = require('./package.json')
const { getAuthenticationHeaders } = require('./lib/sts')
const { log, noop } = require('./lib/log')

const globalHeaders = {
  'X-Cerberus-Client': `CerberusNodeClient/${packageData.version}`
}

const cerberusVersion = 'v1'
const DEFAULT_RETRY_ATTEMPT_NUMBER = 3

/**
 * Options for creating a {@link CerberusClient}
 * @interface CerberusClientOptions
 * @typedef CerberusClientOptions
 * @type {object}
 * @property {string} hostUrl required base url for the Cerberus API.
 * @property {string} [region] region to sign sts auth request for, defaults to us-west-2
 * @property {string} [token] Override the cerberus token. Useful for testing
 * @property {boolean} [debug] If set to true additional logging occurs.
 */

/**
 * @interface ListKeyResult
 * @typedef ListKeyResult
 * @type {object}
 * @property {array<string>} keys
 */

/**
 * @interface ListFileResult
 * @typedef ListFileResult
 * @type {object}
 * @property {boolean} has_next If the result requires pagination
 * @property {string} next_offset The offset to use for the next page
 * @property {number} limit The limit that was used for the results
 * @property {number} offset The offset of the results
 * @property {number} file_count_in_result Number of files in result
 * @property {number} total_file_count Number of total files under path
 * @property {array<SecureFileSummary>} secure_file_summaries
 */

/**
 * @interface SecureFileSummary
 * @typedef SecureFileSummary
 * @type {object}
 * @property {string} sdbox_id The SDB id
 * @property {string} path The path for the file
 * @property {number} size_in_bytes The size in bytes of the file
 * @property {string} name The name of the file
 * @property {string} created_by Who originally uploaded the file
 * @property {string} created_ts ISO 8061 String of when the file was originally uploaded
 * @property {string} last_updated_by Who last updated the file
 * @property {string} last_updated_ts ISO 8061 String of when the file was last updated
 */

/**
 * Cerberus client with CRUD operations for secure data and files.
 *
 * @example
 * var CerberusClient = require('cerberus-node-client')
 *
 * var client = new CerberusClient({
 *   // string, The cerberus URL to use.
 *   hostUrl: YOUR_CERBERUS_HOST,
 *
 *   // string, AWS region, required if authenticating to AWS China, otherwise defaults to us-west-2
 *   region: AWS_REGION,
 *
 *   // boolean, defaults to false. When true will console.log many operations
 *   debug: true,
 *
 *   // This will be used as the cerberus X-Vault-Token if supplied
 *   // OVERRIDDEN by process.env.CERBERUS_TOKEN
 *   // If present, normal authentication with cerberus will be skipped
 *   // You should normally only be using this in testing environments
 *   // When developing locally, it is easier to use process.env.CERBERUS_TOKEN
 *   token: 'Some_Auth_Token'
 * })
 *
 * client.getSecureData('path/to/my/secret').then(secureConfig => {
 *   //do something with config
 * })
 */
class CerberusClient {
  /**
   * @param {CerberusClientOptions} options The options for the Cerberus client.
   */
  constructor (options) {
    if (!options || typeof options !== 'object') {
      throw new Error('options parameter is required')
    }
    this._log = options.debug ? log : noop

    if (options.token) {
      this._log('constructor options token found', options.token)
      this._token = options.token
    } else {
      // Override context with env variables
      const envToken = getEnvironmentVariable(process.env.CERBERUS_TOKEN)
      if (envToken) {
        this._log('environment variable token found', envToken)
        this._token = envToken
      }
    }

    // Validate configuration
    if (typeof options.hostUrl !== 'string') {
      throw new Error('options.hostUrl must be a URL string')
    }

    this._hostUrl = options.hostUrl
    this._region = options.region ? options.region : 'us-west-2'
  }

  /**
   * Fetches secure data.
   *
   * @param {string} path The path for the secure data
   * @return {Promise<object>} A promise that when resolved supplies the secure data
   */
  getSecureData (path) {
    return this._doSecretAction('GET', path, undefined)
  }

  /**
   * Writes secure data.
   *
   * @param {string} path The path for the secure data
   * @param {object} data The secure data
   * @return {Promise<undefined>} A promise that will be resolved when the write is finished
   */
  writeSecureData (path, data) {
    return this._doSecretAction('POST', path, data)
  }

  /**
   * Deletes secure data.
   *
   * @param {string} path The path for the secure data
   * @return {Promise<object>} A promise that will be resolved when the delete is finished
   */
  deleteSecureData (path) {
    return this._doSecretAction('DELETE', path, undefined)
  }

  /**
   * lists the keys under a secure data path.
   *
   * If no keys are present {ListKeyResult} will have an empty array.
   *
   * @param {string} path The path or partial path
   * @return {Promise<ListKeyResult>} A promise that will be resolved when the list is finished supplying the results
   */
  async listPathsForSecureData (path) {
    let res
    try {
      res = await this._doSecretAction('LIST', path, undefined)
    } catch (e) {
      // If no keys under a partial path can be found the API returns a 404, lets convert that to set of empty keys
      if (e.message.includes('status code: 404')) {
        res = { keys: [] }
      } else {
        const msg = 'There was an issue trying to list paths for secure data\nmsg: ' + e.message
        this._log(msg)
        throw new Error(msg)
      }
    }
    return res
  }

  /**
   * lists the files under a path.
   *
   * @param {string} path The path or partial path
   * @return {Promise<ListFileResult>} A promise that will be resolved when the list is finished supplying the {ListFileResult}
   */
  listFile (path) {
    return this._doFileAction('LIST', path, undefined)
  }

  /**
   * Reads the contents of an uploaded file
   *
   * @param {string} path The path the the uploaded file
   * @return {Promise<Buffer|string>} A promise that will be resolved when the file contents have been fetched
   */
  readFile (path) {
    return this._doFileAction('GET', path, undefined)
  }

  /**
   * Uploads a file to a given path
   *
   * @param {string} path The path
   * @param {string|Buffer} data The file buffer or string
   * @return {Promise<object>} A promise that will be resolved when the file contents have been uploaded
   */
  writeFile (path, data) {
    return this._doFileAction('POST', path, data)
  }

  /**
   * deletes an uploaded file
   *
   * @param {string} path The path the the uploaded file
   * @return {Promise<object>} A promise that will be resolved when the file contents have been deleted
   */
  deleteFile (path) {
    return this._doFileAction('DELETE', path, undefined)
  }

  /**
   * Performs an API action against the secret endpoint in Cerberus
   *
   * @param type The type of secret action
   * @param {string} path The secure data path
   * @param body The post for writes
   * @return {Promise<*>} This method returns a promised that when resolved will supply the secure data.
   * @private
   */
  async _doSecretAction (type, path, body) {
    this._log(`Starting ${type} request for ${path}`)
    const token = await this._getToken()

    let url = new URL(urlJoin(cerberusVersion, 'secret', path), this._hostUrl).href
    if (type === 'LIST') {
      url += '?list=true'
    }

    const response = await this._executeCerberusRequest({
      headers: Object.assign({}, globalHeaders, { 'X-Cerberus-Token': token }),
      method: type === 'LIST' ? 'GET' : type,
      url,
      body
    })

    return response ? response.data : undefined
  }

  /**
   * Executes a request against the Cerberus API dealing with any error cases.
   *
   * @param requestConfig The request configuration to be executed
   * @return {Promise<*>} The response body from Cerberus
   * @private
   */
  async _executeCerberusRequest (requestConfig) {
    let response
    try {
      response = await this._executeRequest(Object.assign({}, { json: true }, requestConfig))
      if (!response) {
        throw new Error('No response was returned from Cerberus')
      }
    } catch (error) {
      const msg = 'There was an error executing a call to Cerberus.\nmsg: \'' + error.message + '\''
      this._log(msg)
      throw new Error(msg)
    }

    if (!(response.statusCode >= 200 && response.statusCode < 300)) {
      if (response.headers['content-type'] && response.headers['content-type'].startsWith('application/json')) {
        const msg = `Cerberus returned an error, when executing a call.\nstatus code: ${response.statusCode}\nmsg: ${JSON.stringify(response.data)}`
        this._log(msg)
        throw new Error(msg)
      } else {
        const msg = `Cerberus returned a non-success response that wasn't JSON, this is likely due to being blocked by the WAF.\nstatus code: ${response.statusCode}\nmsg: ${response.data ? response.data : ''}`
        this._log(msg)
        throw new Error(msg)
      }
    }

    return response.data
  }

  // noinspection JSMethodCanBeStatic
  /**
   * Uses the micro request library to execute the request
   * @param requestConfig
   * @return {promise<*>}
   * @private
   */
  async _executeRequest (requestConfig) {
    let resp
    try {
      return await backoff.backOff(async () => {
        resp = await request(requestConfig)
        if (!resp) {
          return resp
        } else if (resp.statusCode < 500) {
          return resp
        } else {
          throw new Error('Cerberus returned a Server Error, executing retry')
        }
      }, { numOfAttempts: DEFAULT_RETRY_ATTEMPT_NUMBER })
    } catch (e) {
      return resp
    }
  }

  /**
   * Uses setTimeout to make a sleep function
   *
   * @param {int} type - the number of milliseconds to sleep for
   * @returns {promise<*>}
   * @private
   */
  _sleep (ms) {
    return new Promise(resolve => setTimeout(resolve, ms))
  }

  /**
   * Upload, delete, read, and list files on Cerberus.
   *
   * @param {string} type - The HTTP method (with the exception of 'LIST') to use as outlined in https://github.com/Nike-Inc/cerberus-management-service/blob/master/API.md
   * @param {string} filePath - The path of the file
   * @param {string|Buffer} fileBuffer - Buffer of the file to upload
   * @returns {Promise<object>} Buffer if read file and Cerberus response otherwise
   * @private
   */
  async _doFileAction (type, filePath, fileBuffer) {
    this._log(`Starting ${type} file request for ${filePath}`)
    const token = await this._getToken()
    let form
    if (fileBuffer) {
      form = new FormData({})
      form.append('file-content', fileBuffer, { filename: filePath.match(/([^/]*)\/*$/)[1] })
    }

    let securePath = 'secure-file'
    if (type === 'LIST') {
      securePath += 's'
    }
    const pathName = urlJoin(cerberusVersion, securePath, filePath)
    const url = new URL(pathName, this._hostUrl).href
    const data = await this._executeCerberusRequest({
      method: type === 'LIST' ? 'GET' : type,
      url,
      headers: Object.assign({}, globalHeaders, { 'X-Cerberus-Token': token }, type === 'POST' ? form.getHeaders() : undefined),
      body: form,
      json: type === 'LIST' || type === 'DELETE'
    })

    return data
  }

  /**
   * Fetches a token either from a local env var or attempts to authenticate with Cerberus via the STS authentication endpoint.
   *
   * @return {Promise<string>} when the promise is resolved the Cerberus auth token string is supplied.
   * @private
   */
  async _getToken () {
    // tokenExpiresAt in secs, Date.now in ms
    if (this._tokenExpiresAt && (this._tokenExpiresAt <= (Date.now() / 1000))) {
      this._tokenExpiresAt = null
      this._token = null
    }

    // Already has token
    if (this._token) {
      this._log('returning stored token')
      return this._token
    }

    let authResponse
    try {
      const authHeaders = await getAuthenticationHeaders(this._region)
      authResponse = await this._executeCerberusRequest({
        method: 'POST',
        url: new URL('v2/auth/sts-identity', this._hostUrl).href,
        headers: Object.assign({}, globalHeaders, authHeaders)
      })
    } catch (error) {
      const msg = 'There was an issue trying to authenticate with Cerberus using the STS auth endpoint\nmsg: \'' + error.message + '\''
      this._log(msg)
      throw new Error(msg)
    }

    if (authResponse.metadata.aws_iam_principal_arn != null) {
      this._log('Successfully authenticated with Cerberus as ' + authResponse.metadata.aws_iam_principal_arn)
    } else if (authResponse.metadata.username != null) {
      this._log('Successfully authenticated with Cerberus as ' + authResponse.metadata.username)
    }
    // Expire 60 seconds before lease is up, to account for latency
    this._tokenExpiresAt = (Date.now() / 1000) + authResponse.lease_duration - 60 // token TTL in secs, Date.now in ms
    this._token = authResponse.client_token
    return this._token
  }
}

/**
 * Gets the set value or undefined
 *
 * @param value The value under question
 * @return {String|undefined} Returns the string of the value or undefined
 * @private
 */
const getEnvironmentVariable = (value) => {
  return value && value !== 'undefined' && value !== undefined && value !== null ? value : undefined
}

module.exports = CerberusClient
